From f522d9d62f1c66e01c4f4156cc5446f4666d580e Mon Sep 17 00:00:00 2001
From: Pawel Moll <pawel.moll@st.com>
Date: Thu, 19 Mar 2009 13:03:01 +0000
Subject: [PATCH] sound_stm: Remove hardware-issue workaround for recent PCM readers

The spurious request issue has been finally fixed in the recent
PCM reader hardware, so we may use the full FDMA transfer size now.
The first SOC with this fix is 7200 cut 3.

Signed-off-by: Pawel Moll <pawel.moll@st.com>
---
 sound/stm/pcm_reader.c    |    9 +++++----
 sound/stm/reg_aud_pcmin.h |    3 +++
 sound/stm/stx7200.c       |    6 ++++--
 3 files changed, 12 insertions(+), 6 deletions(-)

diff --git a/sound/stm/pcm_reader.c b/sound/stm/pcm_reader.c
index 78e6755..abadec5 100644
--- a/sound/stm/pcm_reader.c
+++ b/sound/stm/pcm_reader.c
@@ -382,10 +382,11 @@ static int snd_stm_pcm_reader_hw_params(struct snd_pcm_substream *substream,
 		set__AUD_PCMIN_FMT__DMA_REQ_TRIG_LMT(pcm_reader, transfer_size);
 		set__AUD_PCMIN_FMT__BACK_STALLING__DISABLED(pcm_reader);
 
-		/* This is a workaround for a problem with PCM Reader
-		 * FIFO underrunning (!!!), caused by FDMA issuing
-		 * more than one read per request line assertion... */
-		if (transfer_size > 2)
+		/* This is a workaround for a problem in early releases
+		 * of multi-channel PCM Readers with FIFO underrunning (!!!),
+		 * caused by spurious request line generation... */
+		if (pcm_reader->ver < ver__AUD_PCMIN__65_3_4 &&
+				transfer_size > 2)
 			fdma_req_config.count = transfer_size / 2;
 		else
 			fdma_req_config.count = transfer_size;
diff --git a/sound/stm/reg_aud_pcmin.h b/sound/stm/reg_aud_pcmin.h
index 61ace08..4210967 100644
--- a/sound/stm/reg_aud_pcmin.h
+++ b/sound/stm/reg_aud_pcmin.h
@@ -20,6 +20,9 @@
 /* 7200 2.0 */
 #define ver__AUD_PCMIN__65_3_3 5
 
+/* 7200 3.0 */
+#define ver__AUD_PCMIN__65_3_4 6
+
 
 
 /*
diff --git a/sound/stm/stx7200.c b/sound/stm/stx7200.c
index a2698e7..a499fc0 100644
--- a/sound/stm/stx7200.c
+++ b/sound/stm/stx7200.c
@@ -731,6 +731,7 @@ static int __init snd_stm_stx7200_init(void)
 {
 	int result;
 	struct snd_card *card;
+	int ver;
 
 	snd_stm_printd(0, "snd_stm_stx7200_init()\n");
 
@@ -800,8 +801,9 @@ static int __init snd_stm_stx7200_init(void)
 		hdmi_spdif_player.resource[0].start = 0xfd112c00;
 		hdmi_spdif_player.resource[0].end = 0xfd112c43;
 
-		SET_VER(snd_stm_pcm_reader_info, pcm_reader_0, 5);
-		SET_VER(snd_stm_pcm_reader_info, pcm_reader_1, 5);
+		ver = (cpu_data->cut_major == 2 ? 5 : 6);
+		SET_VER(snd_stm_pcm_reader_info, pcm_reader_0, ver);
+		SET_VER(snd_stm_pcm_reader_info, pcm_reader_1, ver);
 
 		break;
 
-- 
1.6.0.6

